WcaOnRails/app/models/event.rb

```diff
@@ -1,241 +1,32 @@
-class Event
-  attr_accessor :id, :name, :rank, :format, :cellName, :valid, :preferred_formats
-  alias_method :valid?, :valid
+class Event < ActiveRecord::Base
+  self.table_name = "Events"
 
-  def initialize(attributes={})
-    @id = attributes[:id]
-    @name = attributes[:name]
-    @rank = attributes[:rank]
-    @format = attributes[:format]
-    @cellName = attributes[:cellName]
-    @valid = attributes[:valid]
+  has_many :competitions_events
+  has_many :competitions, through: :competition_events
+  has_many :registration_events
+  has_many :registrations, through: :registration_events
+  has_many :user_preferred_events
+  has_many :users, through: :user_preferred_events
+  has_many :preferred_formats
+  has_many :formats, through: :preferred_formats
 
-    # Maps event codes to an array of allowed formats, in decreasing order of
-    # preference. Built from https://www.worldcubeassociation.org/regulations/#9b
-    @preferred_formats = attributes[:preferred_format_ids].map { |format_id| Format.find(format_id) }
-  end
-
-  def to_partial_path
-    "event"
-  end
-
-  def self.find(id)
-    ALL_EVENTS_BY_ID[id] or raise "Unrecognized event id"
-  end
+  scope :all_official, -> { where("rank < 990").order(:rank) }
+  scope :all_deprecated, -> { where("rank between 990 and 999").order(:rank) }
+  scope :all_never_were_official, -> { where("rank >= 1000").order(:rank) }
 
-  def self.find_by_id(id)
-    ALL_EVENTS_BY_ID[id] || Event.new(id: id, name: "Invalid", cellName: "Invalid", rank: 0, valid: false, preferred_format_ids: [])
-  end
-
-  def self.all
-    ALL_EVENTS
+  def recommended_format
+    formats.recommended.first
   end
 
   def official?
-    valid? && rank < 990
+    rank < 990
   end
 
   def deprecated?
-    valid? && 990 <= rank && rank < 1000
+    rank >= 990 && rank < 1000
```

Comment:
Just to throw out another convention, I always like putting the thing that is being "squeezed" in the middle, like so:

```ruby
990 <= rank && rank < 1000
```

This comes from python, who has this awesome construct I love:

```python
990 <= rank < 1000
```

I wish ruby had it =(